daemon/logger/fluentd/fluentd.go

```diff
@@ -0,0 +1,99 @@
+package fluentd
+
+import (
+	"io"
+	"strconv"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/daemon/logger"
+	"github.com/fluent/fluent-logger-golang/fluent"
+)
+
+type Fluentd struct {
+	tag           string
+	containerID   string
+	containerName string
+	writer        *fluent.Fluent
+}
+
+const name = "fluentd"
+
+const defaultHostName = "localhost"
+const defaultPort = 24224
+const defaultTagPrefix = "docker"
+
+func init() {
+	if err := logger.RegisterLogDriver(name, New); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func New(ctx logger.Context) (logger.Logger, error) {
+	config := ctx.Config
+	port := defaultPort
+	host := defaultHostName
+	if config["port"] != "" {
+		portp, errp := strconv.Atoi(config["port"])
+		if errp != nil {
+			return nil, errp
+		}
+		port = portp
+	}
+	if config["host"] != "" {
+		host = config["host"]
+	}
+	tag := config["tag"]
+	if tag == "" {
+		tagPrefix := defaultTagPrefix
+		if config["tag-prefix"] != "" {
+			tagPrefix = config["tag-prefix"]
+		}
+		switch config["tag-type"] {
+		case "name":
+			tag = tagPrefix + "." + ctx.ContainerName
+		case "fullid":
+			tag = tagPrefix + "." + ctx.ContainerID
+		default:
+			tag = tagPrefix + "." + ctx.ContainerID[:12]
+		}
+	}
+	log, err := fluent.New(fluent.Config{FluentPort: port, FluentHost: host})
+	logrus.Debugf("logging driver fluentd configured for container:%s.", ctx.ContainerID)
+	logrus.Debugf("logging driver fluentd port:%d.", port)
+	logrus.Debugf("logging driver fluentd host:%s.", host)
+	logrus.Debugf("logging driver fluentd tag:%s.", tag)
+	if err != nil {
+		return nil, err
+	}
+	return &Fluentd{
+		tag:           tag,
+		containerID:   ctx.ContainerID,
+		containerName: ctx.ContainerName,
+		writer:        log,
+	}, nil
+}
+
+func (f *Fluentd) Log(msg *logger.Message) error {
+	data := map[string]string{
+		"container_id":   f.containerID,
+		"container_name": f.containerName,
+		"source":         msg.Source,
+		"log":            string(msg.Line),
```

Comment:
Tag is for routing in Fluentd world.
And values of `container_id` and `container_name` are needed to be stored into log database, to calculate many things like `avg(count(*)) from logs group by container_id`.
Of course, users can omit any fields which they doesn't need with output configuration.